src/librbd/crypto/Utils.cc

```diff
@@ -30,8 +31,14 @@ void set_crypto(I *image_ctx,
 
   auto crypto = encryption_format->get_crypto();
 
-  auto object_dispatch = CryptoObjectDispatch<I>::create(image_ctx, crypto);
-  image_ctx->io_object_dispatcher->register_dispatch(object_dispatch);
+  if (image_ctx->io_image_dispatcher->exists(
+          io::IMAGE_DISPATCH_LAYER_MIGRATION)) {
+    auto image_dispatch = CryptoImageDispatch<I>::create(image_ctx, crypto);
+    image_ctx->io_image_dispatcher->register_dispatch(image_dispatch);
+  } else {
+    auto object_dispatch = CryptoObjectDispatch<I>::create(image_ctx, crypto);
+    image_ctx->io_object_dispatcher->register_dispatch(object_dispatch);
```

Comment:
I'm may be off because it's been a long time, but:
Why wouldn't the `CryptoImageDispatch` be able to handle `NativeFormat` decryption?

Reply:
It was a suspicion based on the fact that by replacing the object dispatch layer with the new "read-only" image dispatch layer you disable the decryption step in `NativeFormat`, but at the same time still perform the remapping in `align_image_extents()`.

Looking at the code, I guess the question is what type of extents does `FormatInterface` need to be passed.  Because `NativeFormat` implementation assumes that it's passed `DATA` area extents: its `read()` just forwards to the lower layer so it's a bit harder to see, but take a look at `NativeFormat::list_snaps()`:
```
  auto req = io::ImageDispatchSpec::create_list_snaps(
    *m_image_ctx, io::IMAGE_DISPATCH_LAYER_MIGRATION, aio_comp,
    std::move(image_extents), io::ImageArea::DATA, std::move(snap_ids),
    list_snaps_flags, snapshot_delta, {});
  req->send();
```
This meas that, with this PR, an extent would be remapped twice for `NativeFormat`: once in `align_image_extents()` and then as part of normal dispatch on the source image.  Needless to say that that would read garbage data...

Reply:
>  Needless to say that that would read garbage data...

Definitely need to add a test for the live migrating case (`rbd migration prepare src dst` instead of `rbd migration prepare --import-only --source-spec <type + stream spec> dst`).

Reply:
> This meas that, with this PR, an extent would be remapped twice for `NativeFormat`: once in `align_image_extents()` and then as part of normal dispatch on the source image. Needless to say that that would read garbage data...

You're right.
So how about in `CryptoImageDispatch` I re-parse `m_image_ctx->migration_info.source_spec` to get the format, and skip `align_image_extents` if it's a native format.
Another option is to check (again, via parsing migration info) the format at `librbd::crypto::set_crypto` and if it's a native format, then load `CryptoObjectDispatch` instead of the `CryptoImageDispatch` (as was done before this PR).
Your thoughts? Any other idea?

Reply:
I think re-parsing anything that had previously been parsed would be a code smell.  In addition, it would make `FormatInterface` interface be inconsistent and harder to reason about in the future because one implementation would be expecting logical extents and others raw extents.

I would look in the direction of doing the remapping in the individual format implementations, only for the formats where it's needed.  Since dispatch is operating on logical extents, it makes sense to stick to logical extents at the `FormatInterface` boundary IMO:
- `CryptoImageDispatch` would perform alignment and decryption but not remapping; `align_image_extents()` would cease being a misnomer.
- This should fit into the current "remapping is unconditional, encryption/decryption is conditional on whether the appropriate dispatch layer is loaded" scheme.

Reply:
> Definitely need to add a test for the live migrating case (`rbd migration prepare src dst` instead of `rbd migration prepare --import-only --source-spec <type + stream spec> dst`).

Also consider and add a sub-test for a case where `src` is a half-populated cloned image, with the parent being encrypted differently.

Reply:
> I would look in the direction of doing the remapping in the individual format implementations, only for the formats where it's needed.

I see only Native, Raw, and QCOW formats.
Obviously I will add offset remapping to RawFormat.
Regarding QCOW, I think it is currently possible for a user to use LUKS format on-top of qcow.
So I guess I should copy the remap code I plan to add to RawFormat to QCOWFormat as well?

Reply:
> So I guess I should copy the remap code I plan to add to RawFormat to QCOWFormat as well?

Yup.  Built-in LUKS encryption in QCOW2 would have been an additional complication, but `QCOWFormat` doesn't support it so we don't have to worry about it right now.

Reply:
You missed missed my

> Also consider and add a sub-test for a case where `src` is a half-populated cloned image, with the parent being encrypted differently.

suggestion and unfortunately this is still broken in the sense that this PR introduces a regression for the native format.  The issue is that when a half-populated (or completely unpopulated) clone is being migrated, the parent data is decrypted in the appropriate parent (using that parent's key).  Once the parent data reaches the "source" clone -- the one that has `CryptoImageDispatch` loaded -- it's run through decryption again using the clone's key despite already being plaintext.

I think we are going to need to either move decryption into the migration layer and get rid of `CryptoImageDispatch` altogether since it doesn't fit into the dispatch layers concept well or stick with `CryptoImageDispatch` and figure out a passable approach to load it only for non-native formats, as you alluded to in https://github.com/ceph/ceph/pull/44366#discussion_r1325694988.

And improve test coverage, of course -- it shouldn't be me discovering these bugs while play testing.